designs/helm-chart-autogenerate-plugin.md

```diff
@@ -0,0 +1,287 @@
+
+To achieve this goal, we are proposing a new native Kubebuilder [plugin](https://book.kubebuilder.io/plugins/plugins) (i.e., helm-chart/v1-alpha) which will provide the necessary scaffolds. The plugin will function similarly to the existing [Grafana Plugin](https://book.kubebuilder.io/plugins/grafana-v1-alpha), generating or regenerating HelmChart files using the init and edit sub-commands (i.e., `kubebuilder init|edit --plugins helm-chart/v1-alpha`). An alternative solution could be to implement an alpha command, similar to the [helper provided to upgrade projects](https://book.kubebuilder.io/reference/rescaffold).
+
+## Example
+
+**To enable the helm-chart generation when a project is initialized**
+
+> kubebuilder init --plugins=`go/v4,helm-chart/v1-alpha`
+
+**To enable the helm-chart generation after the project be scaffolded OR to update the helm-chart files**
+
+> kubebuilder edit --plugins=`go/v4,helm-chart/v1-alpha`
+
+#### Case 1: Initial Helm chart distribution
+
+For users who have developed an Kubebuilder Project and wish to distribute it via a Helm chart for the first time:
+
+By running `kubebuilder init|edit --plugins=helm-chart/v1alpha`, the scaffolds of Helm Chart to package the Project would be created in the project as follows:
+
+```shell
+$ tree
+.
+..
+├── PROJECT
+...
+├── deploy
+│   └── helm-chart // The helm chart with all values will be scaffold under a directory of the project
+...
+```
+
+The initial setup will be straightforward, creating the Helm chart structure and populating it with the default values as determined by the project's configuration.
+
+#### Case 2: Update the Helm chart with manifests changes
+
+For users who want to update an existing Helm chart to reflect changes in the kustomize directory, such as CRDs, Webhooks, or RBAC roles:
+
+The plugin will automates this process. When you run the plugin, it will generate the necessary manifests and synchronize them with the Helm chart. This ensures that all changes in your kubebuilder project files, such as CRDs, RBAC roles, and webhooks, are accurately reflected in the Helm chart.
+
+#### Case 3: Updating the Helm Chart with Customizations
+
+For users who have customized their Helm chart (e.g., added new configurations in `values.yaml`) and want to update it while preserving these customizations:
+
+Just like the other plugins in Kubebuilder, the Helm Chart plugin will not re-write the `values.yaml` after the initial generation. This means that any customizations made by the user will be preserved.
+
+## Open Questions [optional]
+
+- How to effectively manage the lifecycle of CRDs within the generated Helm chart to support upgrades and changes to CRDs while minimizing potential impacts on existing resources and user deployment processes?
+- How to deal with CertManager and Prometheus? Should we add them as dependencies to the Helm Chart?
```

Comment:
```suggestion
## Open Questions

- How to manage and scaffold the CRDs for the HelmChart?

According to [Helm Best Practices for Custom Resource Definitions](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-1-let-helm-do-it-for-you), 
there are two main methods for handling CRDs:

- **Method 1:Let Helm Do It For You:**  Place CRDs in the `crds/` directory. Helm installs these CRDs during the initial 
install but does not manage upgrades or deletions.
- **Method 2:Separate Charts:**  Place the CRD definition in one chart 
and the resources using the CRD in another chart. 
This method requires separate installations for each chart.

**Raised Considerations and Concerns**
- **Use Helm crd directory** The upgraded chart versions will silently ignore CRDs even if they differ from the installed versions. This could lead to surprising and unexpected behavior.
However, it is how [cert-manager][]
- **Templates Folder**: Moving CRDs to the `templates` folder facilitates upgrades but uninstalls CRDs when the operator is uninstalled.
- **Separate Helm Chart for CRDs:** This approach allows control over both CRD and operator versions without deleting CRDs when the operator chart is deleted.
- **When Webhooks are used:** If a CRD specifies, for example, a conversion webhook, the "API chart" needs to contain the CRDs and the webhook `service/workload`. 
It would also make sense to include `validating/mutating` webhooks, requiring the scaffolding of separate main modules and image builds for 
webhooks and controllers which does not shows to be compatible with Kubebuilder Golang scaffold.

**Proposed Solution**

Follow the same approach adopted by [Cert-Manager](https://cert-manager.io/docs/installation/helm/).
Add the CRDs under the `template` directory and have a spec in the `values.yaml`
which will define if the CRDs should or not be applied

```shell
helm install|upgrade \
  myrelease \
  --namespace my-namespace \
  --set `crds.enabled=true`
```

Additionally, we might want to consider in the future also
scaffold separate charts for the APIs and support both.
An example of this approach provided as feedback
was [karpenter-provider-aws](https://github.com/aws/karpenter-provider-aws/tree/main/charts).

- How to manage dependencies such as Cert-Manager and Prometheus?

Helm charts allow maintainers to define dependencies via `requirements.yaml`.

However, we can convey that in the initial version of this plugin, we do not need to consider any dependencies. 
As it stands today, consumers of the solutions built with Kubebuilder 
are responsible for installing their dependencies prior 
to installing the project built with Kubebuilder.

For example:
- If the project is using cert-manager, or
- If managing metrics is required, ensure that the cluster has Prometheus installed and with the required permissions beforehand.

It seems out of the scope of the project itself to manage these dependencies.

```

Reply:
@dashanji  we should track our discussion and conveys so far.